lib/backend/etcdbk/etcd.go

```diff
@@ -649,6 +649,85 @@ func (b *EtcdBackend) NewWatcher(ctx context.Context, watch backend.Watch) (back
 	return b.buf.NewWatcher(ctx, watch)
 }
 
+func (b *EtcdBackend) Items(ctx context.Context, params backend.IterateParams) iter.Seq2[backend.Item, error] {
+	if params.StartKey.IsZero() {
+		err := trace.BadParameter("missing parameter startKey")
+		return func(yield func(backend.Item, error) bool) { yield(backend.Item{}, err) }
+	}
+	if params.EndKey.IsZero() {
+		err := trace.BadParameter("missing parameter endKey")
+		return func(yield func(backend.Item, error) bool) { yield(backend.Item{}, err) }
+	}
+
+	sort := clientv3.SortAscend
+	if params.Descending {
+		sort = clientv3.SortDescend
+	}
+
+	const defaultPageSize = 1000
+	return func(yield func(backend.Item, error) bool) {
+		inclusiveStartKey := b.prependPrefix(params.StartKey)
+		count := 0
+
+		pageSize := defaultPageSize
+		for {
+			if params.Limit > backend.NoLimit {
+				pageSize = min(params.Limit-count, defaultPageSize)
+			}
+
+			start := b.clock.Now()
+			re, err := b.clients.Next().Get(ctx, inclusiveStartKey,
+				// etcd's range query includes the start point and excludes the end point,
+				// but Backend.GetRange is supposed to be inclusive at both ends, so we
+				// query until the very next key in lexicographic order (i.e., the same key
+				// followed by a 0 byte)
+				clientv3.WithRange(b.prependPrefix(params.EndKey)+"\x00"),
+				clientv3.WithSort(clientv3.SortByKey, sort),
+				clientv3.WithLimit(int64(pageSize)),
+			)
+			batchReadLatencies.Observe(time.Since(start).Seconds())
+			batchReadRequests.Inc()
+			if err := convertErr(err); err != nil {
+				if !yield(backend.Item{}, trace.Wrap(err)) {
+					return
+				}
```

Comment:
We shouldn't be potentially continuing after yielding an error.  I think yielding an error should result in termination of the iterator.

In fact, I wouldn't be opposed to panicing if `yield` returns true after a call with an error value.  Seems like that would indicate a serious logic bug in the caller.

(comment applies to a number of locations in this code)

Reply:
I disagree on panicking or logging an error when iterating "past the end"; it might be a little contrived, but it's not unreasonable to want to zip or merge two fallible Seq2s, maybe with one of [the proposed x/exp/iter adapters](https://github.com/golang/go/issues/61898), and continuing the iteration while collecting errors to aggregate them at a later point seems somewhat reasonable IMO.